pkg/certificate/mrc_reconciler.go

```diff
@@ -1,214 +1,176 @@
+	desiredSigningIssuer, err := m.getCertIssuer(desiredSigningMRC)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	// If it's not in a terminal state, set the next transition time.
-	// NOTE: consider that this isTerminal check will need to apply against the updated mrc state which should happen
-	// above in this method.
-	if isTerminal(mrc) {
-		mrc.Status.TransitionAfter = nil
-	} else {
-		mrc.Status.TransitionAfter = &metav1.Time{Time: time.Now().Add(mrcDurationPerStage)}
+	if desiredSigningMRC == desiredValidatingMRC {
+		desiredValidatingIssuer = desiredSigningIssuer
+		return desiredSigningIssuer, desiredValidatingIssuer, nil
 	}
 
-	// TODO(5046): add the retry loop.
-	return m.mrcClient.UpdateMeshRootCertificate(mrc)
-}
+	desiredValidatingIssuer, err = m.getCertIssuer(desiredValidatingMRC)
+	if err != nil {
+		return nil, nil, err
+	}
 
-func (m *Manager) shouldSetIssuers(mrc *v1alpha2.MeshRootCertificate) bool {
-	// TODO(5046): check the states, and if in some form of an active state, AND if the MRC is not already the existing
-	// then we should set MRC, so return true
-	return true
+	return desiredSigningIssuer, desiredValidatingIssuer, nil
 }
 
-func (m *Manager) setIssuers(mrc *v1alpha2.MeshRootCertificate) error {
+func (m *Manager) getCertIssuer(mrc *v1alpha2.MeshRootCertificate) (*issuer, error) {
+	m.mu.Lock()
+	signingIssuer := m.signingIssuer
+	validatingIssuer := m.validatingIssuer
+	m.mu.Unlock()
+
+	// if the issuer has already been created for the specified MRC,
+	// return the existing issuer
+	if signingIssuer != nil && mrc.Name == signingIssuer.ID {
+		return signingIssuer, nil
+	}
+	if validatingIssuer != nil && mrc.Name == validatingIssuer.ID {
+		return validatingIssuer, nil
+	}
+
 	client, ca, err := m.mrcClient.GetCertIssuerForMRC(mrc)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	c := &issuer{Issuer: client, ID: mrc.Name, CertificateAuthority: ca, TrustDomain: mrc.Spec.TrustDomain, SpiffeEnabled: mrc.Spec.SpiffeEnabled}
-	switch {
-	case mrc.Status.State == constants.MRCStateActive:
-		m.mu.Lock()
-		m.signingIssuer = c
-		m.validatingIssuer = c
-		m.mu.Unlock()
-	case mrc.Status.State == constants.MRCStateIssuingRollback || mrc.Status.State == constants.MRCStateIssuingRollout:
-		m.mu.Lock()
-		m.signingIssuer = c
-		m.mu.Unlock()
-	case mrc.Status.State == constants.MRCStateValidatingRollback || mrc.Status.State == constants.MRCStateValidatingRollout:
-		m.mu.Lock()
-		m.validatingIssuer = c
-		m.mu.Unlock()
-		// TODO(5046): let's not have a default case.. what are the explicit states that can result in this?
-	default:
-		m.mu.Lock()
-		m.signingIssuer = c
-		m.validatingIssuer = c
-		m.mu.Unlock()
+	return c, nil
+}
+
+func filterOutInactiveMRCs(mrcList []*v1alpha2.MeshRootCertificate) []*v1alpha2.MeshRootCertificate {
+	n := 0
+	for _, mrc := range mrcList {
+		if mrc.Spec.Intent != v1alpha2.InactiveIntent {
+			mrcList[n] = mrc
+			n++
+		}
 	}
-	return nil
+	return mrcList[:n]
```

Comment:
is this protecting against potentially 3 active MRC's? That seems like a critical issue and we should return an error?